tests/test_discovery.py

```diff
@@ -533,14 +538,148 @@ def test_api_endpoint_override_from_client_options_dict(self):
+
+    def test_mtls_not_trigger_if_http_provided(self):
+        discovery = open(datafile("plus.json")).read()
+        plus = build_from_document(discovery, http=httplib2.Http())
+        self.assertTrue(plus is not None)
+        self.assertEqual(plus._baseUrl, self.REGULAR_ENDPOINT)
+        self.check_http_client_cert(plus, has_client_cert=False)
+
+    def test_exception_with_client_cert_source(self):
+        discovery = open(datafile("plus.json")).read()
+        with self.assertRaises(MutualTLSChannelError):
+            build_from_document(
+                discovery,
+                credentials=self.MOCK_CREDENTIALS,
+                client_options={"client_cert_source": mock.Mock()},
+            )
+
+    @parameterized.expand(["Never", "Auto", "Always"])
+    def test_mtls_with_provided_client_cert(self, use_client_cert):
+        discovery = open(datafile("plus.json")).read()
+        os.environ["GOOGLE_API_USE_MTLS"] = use_client_cert
+
+        plus = build_from_document(
+            discovery,
+            credentials=self.MOCK_CREDENTIALS,
+            client_options={
+                "client_encrypted_cert_source": self.client_encrypted_cert_source
+            },
+        )
+        self.assertTrue(plus is not None)
+        self.check_http_client_cert(plus, has_client_cert=True)
+        if use_client_cert != "Never":
+            self.assertEqual(plus._baseUrl, self.MTLS_ENDPOINT)
+        else:
+            self.assertEqual(plus._baseUrl, self.REGULAR_ENDPOINT)
+
+        del os.environ["GOOGLE_API_USE_MTLS"]
+
+    @parameterized.expand(["Never", "Auto", "Always"])
+    def test_endpoint_not_switch(self, use_client_cert):
+        # Test endpoint is not switched if user provided api endpoint
+        discovery = open(datafile("plus.json")).read()
+        os.environ["GOOGLE_API_USE_MTLS"] = use_client_cert
+
+        plus = build_from_document(
+            discovery,
+            credentials=self.MOCK_CREDENTIALS,
+            client_options={
+                "api_endpoint": "https://foo.googleapis.com",
+                "client_encrypted_cert_source": self.client_encrypted_cert_source,
+            },
+        )
+        self.assertTrue(plus is not None)
+        self.check_http_client_cert(plus, has_client_cert=True)
+        self.assertEqual(plus._baseUrl, "https://foo.googleapis.com")
+
+        del os.environ["GOOGLE_API_USE_MTLS"]
+
+    @parameterized.expand(["Never", "Auto", "Always"])
+    def test_mtls_with_default_client_cert(self, use_client_cert):
+        discovery = open(datafile("plus.json")).read()
+        os.environ["GOOGLE_API_USE_MTLS"] = use_client_cert
+
+        with mock.patch(
+            "google.auth.transport.mtls.has_default_client_cert_source",
+            return_value=True,
+        ):
+            with mock.patch(
+                "google.auth.transport.mtls.default_client_encrypted_cert_source",
+                return_value=self.client_encrypted_cert_source,
+            ):
```

Comment:
I think this would be easier to read with the decorator syntax for `patch()`

Reply:
done